File: 725. Split Linked List in Parts.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def splitListToParts(self, root: ListNode, k: int) -> List[ListNode]:
        length = self.lengthofll(root)
        nodenum, addonenum = divmod(length, k)
        result = []
        addcount = 0
        prev = None
        for i in range(k):
            dmhead = ListNode(0)
            result.append(dmhead)
            dmhead.next = root
            for j in range(nodenum):
                prev = root
                root = root.next
            if addcount < addonenum:
                prev = root
                root = root.next
                addcount += 1
            if prev:
                prev.next = None
        return [i.next for i in result]
        
    def lengthofll(self, head):
        count = 0
        while head:
            count += 1
            head = head.next
        return count

File: 725. Split Linked List in Parts.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode[] splitListToParts(ListNode root, int k) {
        ListNode node = root;
        ListNode[] res = new ListNode[k];
        int len = 0;
        while (node != null) {
            len++;
            node = node.next;
        }
        if (len == 0) {
            return res;
        }
        
        int partlen = len/k, count;
        node = root;
        ListNode prev = null;
        for (int i=0; i<k; i++) {
            count = 0;
            if (node == null) {
                break;
            }
            res[i] = node;
            while (count < partlen) {
                prev = node;
                node = node.next;
                count++;
            }
            if (i < len%k) {
                prev = node;
                node = node.next;
            }
            prev.next = null;
        }
        return res;
    }
}